### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by MartinWohlan**
- Added testing to the docs
- Added .DS_Store to the gitignore

### CODE REVIEW

File: docs/source/tests/pytest.rst [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        # Arrange
+        data = XY(example_xy_data_pos_int)
+
+        # Act
+        x_position = data.x
+
+        # Assert
+        assert np.array_equal(x_position, np.array([[1, 3], [5, 7]]))
+
+Fixtures are a quite powerful tool since they are modular and can also request other fixtures.
```
Comment (draabe) [inline docs/source/tests/pytest.rst:76]: Maybe one more sentence here for how we use them: to get minimal examples of e.g. data-level objects such as XY, Events, or Code. Of course, they look different depending on the method tested

File: docs/source/tests/pytest.rst [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    #. Inside the testing files.
+    #. Inside a ``conftest.py`` file.
+    #. Inside an extra file which is then integrated into the ``conftest.py`` file as a plugin.
+
+The ``conftest.py`` file just follows a naming convention of pytest and enables to share fixtures across multiple files. The fixtures implemented inside the ``conftest.py`` file can be accessed from testing files laying in the same folder layer or in a subdirectory without any import. For more detailed information (especially on option 3.) have a look on this `link <https://docs.pytest.org/en/6.2.x/fixture.html>`_.
+
+Marks
+------
+Marks can be used to categorize your tests. To do so you need to decorate the method with ``@pytest.mark.<mark_name>``. When executing the ``pytest -m <mark_name>`` command (:doc:`see here </tests/testing>`) only methods decorated with ``@pytest.mark.<mark_name>`` will be selected for the testing. This can be advantageous if you have tests that are slower because they are for example accessing a database but you want to quickly run your test suite.
```
Comment (draabe) [inline docs/source/tests/pytest.rst:94]: The `see here` link will break if joined into one file

File: docs/source/tests/pytest.rst [hunk 3]
```diff
@@ -0,0 +1,124 @@
+============================
+About The Pytest Framework
+============================
+The pytest framework provides a feature-rich, plugin-based ecosystem that helps to easily write small as well as readable tests and it can also scale to support complex functional testing. To make sure that you can use the full functionality of pytest this section provides you some conventions and commands that are useful. If you want to get more into the whole framework you can find further information `here <https://docs.pytest.org/en/6.2.x/contents.html#toc>`__.
+As described in the :doc:`previous section </tests/testing>` pytest follows a strict naming convention for files (``test_*.py``) and methods (``def test_*()``).
```
Comment (draabe) [inline docs/source/tests/pytest.rst:5]: The `previous section` link will break if joined into one file

File: docs/source/tests/pytest.rst [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        assert np.array_equal(x_position, np.array([[1, 3], [5, 7]]))
+
+Pytest comes with a few marks out of the box which can bee seen `here <https://docs.pytest.org/en/6.2.x/mark.html#>`_. To create your own customized mark you have add the following plugin to the ``pyproject.toml`` file:
+
+.. code-block::
+
+    [tool.pytest.ini_options]
+    markers = [
+        "<mark_name1>: description",
+        "<mark_name12: description"
```
Comment (draabe) [inline docs/source/tests/pytest.rst:116]:
```suggestion
        "<mark_name2: description"
```

File: docs/source/tests/pytest.rst [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    ]
+
+Testing workflow
+================
+    #. Before starting the coding session :doc:`run pytest </tests/testing>` in your terminal to see if everything works or you get some errors which have to be fixed.
```
Comment (draabe) [inline docs/source/tests/pytest.rst:121]: The `run pytest` link will break if joined into one file

File: docs/source/tests/pytest.rst [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    #. After or before writing a class or method write the according tests and fixtures to keep your test suite always up to date.
+    #. After finishing your coding session :doc:`run pytest </tests/testing>` again.
```
Comment (draabe) [inline docs/source/tests/pytest.rst:123]: The `run pytest link will break if joined into one file

**Review by draabe — CHANGES_REQUESTED** [review]
Hi @MartinWohlan! Thanks for this great summary on testing. I think its concise, descriptive (with good examples) and well written.

I've added very few comments, but I'd ask you to refactor the files the following:
- Remove the .gitignore files to avoid merge conflicts. There also is no separate .gitignore in subfolders
- Please include both .rst files that contain your documentation as chapters into the "Contributing.rst". I think its best if we keep all docs on how to contribute (and I'd say that includes testing) in a single document. New user's will appreciate that they have a single file that (hopefully) fits all their questions they have when they start to develop. For experienced users, however, all this is not new and so we keep our directories and toctrees clean.
- Once you're finished, you don't need any changes to the top-level file in terms of toctrees

I've commented on all sections where this refactor will break links in the text.

Thanks a lot!

**Comment by MartinWohlan** [comment]
> Hi @MartinWohlan! Thanks for this great summary on testing. I think its concise, descriptive (with good examples) and well written.
> 
> I've added very few comments, but I'd ask you to refactor the files the following:
> 
> * Remove the .gitignore files to avoid merge conflicts. There also is no separate .gitignore in subfolders
> * Please include both .rst files that contain your documentation as chapters into the "Contributing.rst". I think its best if we keep all docs on how to contribute (and I'd say that includes testing) in a single document. New user's will appreciate that they have a single file that (hopefully) fits all their questions they have when they start to develop. For experienced users, however, all this is not new and so we keep our directories and toctrees clean.
> * Once you're finished, you don't need any changes to the top-level file in terms of toctrees
> 
> I've commented on all sections where this refactor will break links in the text.
> 
> Thanks a lot!

**Comment by MartinWohlan** [comment]
> Hi @MartinWohlan! Thanks for this great summary on testing. I think its concise, descriptive (with good examples) and well written.
> 
> I've added very few comments, but I'd ask you to refactor the files the following:
> 
> * Remove the .gitignore files to avoid merge conflicts. There also is no separate .gitignore in subfolders

I wanted to automatically add the .DS_Store file to the .gitignore but was in the wrong directory, so it created a new one. I have deleted the unnecessary files.

File: docs/source/tests/pytest.rst [hunk 7]
```diff
@@ -0,0 +1,124 @@
+============================
+About The Pytest Framework
+============================
+The pytest framework provides a feature-rich, plugin-based ecosystem that helps to easily write small as well as readable tests and it can also scale to support complex functional testing. To make sure that you can use the full functionality of pytest this section provides you some conventions and commands that are useful. If you want to get more into the whole framework you can find further information `here <https://docs.pytest.org/en/6.2.x/contents.html#toc>`__.
+As described in the :doc:`previous section </tests/testing>` pytest follows a strict naming convention for files (``test_*.py``) and methods (``def test_*()``).
+
+How to execute pytest
+=====================
+As part of the `continuous integration <https://www.atlassian.com/continuous-delivery/continuous-integration>`_ pipeline build into the floodlight repository all the tests are going to be executed when making the pull request. Irrespective of this, tests should be carried out internally on a regular basis.
```
[the remark this answers is not in the document]
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:9]: deleted the link

File: docs/source/tests/pytest.rst [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+.. code-block:: shell
+
+    $ pytest -k "<string1> and not <string2>" # to run all tests which contain the <string1> and not the <string2> expression
+
+.. code-block:: shell
+
+    $ pytest <filename>.py::<methode_name> # to run a specific test (<method_name>) within a module (<filename>)
+
+# Should I exclude this examples and just reference the part of the documentation of pytest which contains all these commands
```
[the remark this answers is not in the document]
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:37]: Deleted the comment

File: docs/source/tests/pytest.rst [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+Useful features
+===============
+
+Pytest provides multiple features that are very useful to simplify the testing procedure and keep your tests organized and structured.
+
+Fixtures
+---------
+Most of the tests depend on some sort of input. With `fixtures <https://docs.pytest.org/en/6.2.x/fixture.html>`_ pytest provides a feature with which data, test doubles or some system state can be created. Fixtures are reusable and can be used for multiple tests. In order to create a fixture you have to build a function that returns the data or system state that is needed for your testing. To do that just decorate this function with ``@pytest.fixture``. The function name can now get passed to a testing method as an argument. Here you can see an example of how fixtures can be implemented:
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:51]: Updated

File: docs/source/tests/pytest.rst (hunk 1 above)
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:76]: Added: In a nutshell they can be understood as minimal examples of e.g. data-level objects such as XY, Events, or Code. But compared to the normal objects, they are much clearer and are still able to test the full functionality of the methods. Of course, they look different depending on the method tested.

File: docs/source/tests/pytest.rst (hunk 2 above)
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:94]: all :doc: links where changed to :refs:

File: docs/source/tests/pytest.rst (hunk 3 above)
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:5]: all :doc: links where changed to :refs:

File: docs/source/tests/pytest.rst (hunk 4 above)
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:116]: changed to mark name 2

File: docs/source/tests/pytest.rst (hunk 5 above)
Comment (MartinWohlan) [reply docs/source/tests/pytest.rst:121]: all :doc: links where changed to :refs:

**Review by draabe — APPROVED** [review]

**Comment by draabe** [comment]
Hi, thanks for the changes! Looks all good to me.

On a side note: Please refrain from merging unrelated branches that are not necessarily needed. This clutters the PR and leads to confusing version trees (or even unnecessary merge conflicts). It's okay this time, I'll squash commit the PR and there is merge back to develop scheduled soon.

Thanks again for your addition to the docs!

**Pull request merged**